### ISSUE TITLE
Custom Policy not Inherited in Model

### ISSUE THREAD

**Issue by James-R-Han** [DLR-RM/stable-baselines3#1105]
I am trying to define a custom feature extractor and policy architecture for a DQN. I am also using a custom environment with gym.spaces.Dict as the observation space.
 
My model enters only the __init__ of CustomQ and CustomExtractor (twice for this class), but never enters the forward part of the function.

My model definition is: ```model = DQN(CustomQ, env, policy_kwargs=dict(features_extractor_class=CustomExtractor), verbose=1)```

The main classes I use look like:
```
class CustomQ(MultiInputPolicy):
        def __init__(
            self,
            observation_space: gym.spaces.Space,
            action_space: gym.spaces.Space,
            lr_schedule: Callable[[float], float],
            net_arch: Optional[List[Union[int, Dict[str, List[int]]]]] = None,
            activation_fn: Type[nn.Module] = nn.Tanh,
            *args,
            **kwargs,
        ):

        super(CustomQ, self).__init__(
            observation_space,
            action_space,
            lr_schedule,
            net_arch,
            activation_fn,
            *args,
            **kwargs,
        )
        # Disable orthogonal initialization
        self.ortho_init = False

    def _build_mlp_extractor(self) -> None:
        print("building mlp extractor")
        self.mlp_extractor = CustomQNetwork(self.features_dim)

class CustomQNetwork(nn.Module):
    def __init__(
        self,
        feature_dim: int,
        last_layer_dim_pi: int = 26,
        last_layer_dim_vf: int = 1,
    ):
        super(CustomQNetwork, self).__init__()
        self.latent_dim_pi = last_layer_dim_pi
        self.latent_dim_vf = last_layer_dim_vf
        # I know this isn't a Q Network, but it never even enters the forward of this class and this is working code in the SB3 custom policy documentation
        # Policy network
        self.policy_net = nn.Sequential(
            nn.Linear(feature_dim, last_layer_dim_pi), nn.ReLU()
        )
        # Value network
        self.value_net = nn.Sequential(
            nn.Linear(feature_dim, last_layer_dim_vf), nn.ReLU()
        )

    def forward(self, features) -> Tuple[torch.Tensor, torch.Tensor]:
        """
        :return: (torch.Tensor, torch.Tensor) latent_policy, latent_value of the specified network.
            If all layers are shared, then ``latent_policy == latent_value``
        """
        return self.policy_net(features), self.value_net(features)

    def forward_actor(self, features: torch.Tensor) -> torch.Tensor:
        return self.policy_net(features)

    def forward_critic(self, features: torch.Tensor) -> torch.Tensor:
        return self.value_net(features)

class CustomExtractor(BaseFeaturesExtractor):
    def __init__(self, observation_space: gym.spaces.Dict):
        super(CustomExtractor, self).__init__(observation_space, features_dim=1)

        total_concat_size = 0

        for key, _ in observation_space.spaces.items():
            if "type1" in key:
                total_concat_size += 5                
            else:
                total_concat_size += 9

        self._features_dim = total_concat_size

    def forward(self, observations):
        # for now I was just returning a dictionary of tensors to see if I was allowed, but I never actually enter this forward...
        return observations
```

My custom environment passes the env checker.
And when I run model.learn(total_timesteps=500) it doesn't throw any error even though I know it should as I was just testing it to debug through the errors for what I want.

![image](https://user-images.githubusercontent.com/50814323/194954127-3dcf922f-06f2-4e23-acf3-9a819d58a979.png)

I'm not sure if I'm missing anything obvious, but any help would be greatly appreciated :)

**Comment by araffin** [comment]
Hello,
did you try setting the "learning starts" to a lower value ?
500 steps is very little compared to the default value, my guess is that the forward method is not called because it is still in the warmup phase, with pure exploration.
And please use markdown codeblock to format your code as shown in the issue template.

**Comment by James-R-Han** [comment]
Hi araffin! Thanks for the speedy response.

Yay! That triggered the forward in my feature extractor! 

Would you happen to have the details on how the timesteps before learning_starts works? Where is it getting the actions from in the episodes leading up to that parameter number? Is it just random sampling?

**Comment by qgallouedec** [comment]
Yes, before `learning_starts`, the action is sampled randomly from the action space.
